File: app/controllers/appearance_controller.rb
class AppearanceController < UIViewController
  extend IB

  outlet :previewView, UIView
  outlet :previewImageView, UIImageView
  outlet :previewLabel, UILabel

  outlet :borderView, UIView
  outlet :settingsView, UIView

  outlet :fontView, UIView
  outlet :fontScaleSlider, UISlider
  outlet :fontWeightButton, UIButton

  outlet :backgroundView, UIView

  outlet :fontBackgroundControl, UISegmentedControl
  outlet :colorPickerBorder, UIView

  def viewDidLoad
    super

    @appearance = Appearance.sharedClient
    @previewLabel.text = "#{(CountdownTimer.sharedClient.length / 60).to_s}:#{(CountdownTimer.sharedClient.length % 60).to_s.rjust(2,'0')}"

    @borderView.backgroundColor = UIColor.clearColor
    @borderView.layer.cornerRadius = 6.0
    @borderView.layer.masksToBounds = true
    @borderView.layer.borderWidth = 1.0
    @borderView.layer.borderColor = UIColor.lightGrayColor.CGColor

    NSBundle.mainBundle.loadNibNamed("FontAppearanceView", owner:self, options:nil)
    @fontScaleSlider.value = @appearance.fontScale
    @borderView.addSubview(@fontView)
    layoutView(@fontView, toMatchView: @settingsView)

    NSBundle.mainBundle.loadNibNamed("BackgroundAppearanceView", owner:self, options:nil)
    @backgroundView.alpha = 0
    @borderView.addSubview(@backgroundView)
    layoutView(@backgroundView, toMatchView: @settingsView)

    @colorPickerBorder.backgroundColor = UIColor.clearColor
    @colorPickerBorder.layer.cornerRadius = 6.0
    @colorPickerBorder.layer.masksToBounds = true
    @colorPickerBorder.layer.borderWidth = 1.0
    @colorPickerBorder.layer.borderColor = UIColor.darkGrayColor.CGColor

    @colorPicker = STColorPicker.alloc.initWithFrame(@colorPickerBorder.bounds)
    @colorPicker.setColorHasChanged(lambda { |colorPointer, location|
      @appearance.send("#{@pickColorOf}=", colorPointer.to_object)# = color
      updateAppearance
    })

    @colorPickerBorder.addSubview(@colorPicker)

    @pickColorOf = "textColor"

    updateAppearance

  end

  def layoutView(view1, toMatchView: view2)

    view1.translatesAutoresizingMaskIntoConstraints = false
    view1.superview.addConstraint(NSLayoutConstraint.constraintWithItem(view1, attribute: NSLayoutAttributeLeading, relatedBy: NSLayoutRelationEqual, toItem: view2, attribute: NSLayoutAttributeLeading, multiplier: 1, constant: 0))
    view1.superview.addConstraint(NSLayoutConstraint.constraintWithItem(view1, attribute: NSLayoutAttributeTrailing, relatedBy: NSLayoutRelationEqual, toItem: view2, attribute: NSLayoutAttributeTrailing, multiplier: 1, constant: 0))
    view1.superview.addConstraint(NSLayoutConstraint.constraintWithItem(view1, attribute: NSLayoutAttributeTop, relatedBy: NSLayoutRelationEqual, toItem: view2, attribute: NSLayoutAttributeTop, multiplier: 1, constant: 0))
    view1.superview.addConstraint(NSLayoutConstraint.constraintWithItem(view1, attribute: NSLayoutAttributeBottom, relatedBy: NSLayoutRelationEqual, toItem: view2, attribute: NSLayoutAttributeBottom, multiplier: 1, constant: 0))

  end

  def viewDidLayoutSubviews

    @colorPicker.frame = @colorPickerBorder.bounds

  end

  def done

    presentingViewController.dismissModalViewControllerAnimated(true)

  end

  def reset

    @appearance.reset
    @fontScaleSlider.value = @appearance.fontScale
    
    updateAppearance

  end

  def chooseBackground

    imagePicker = UIImagePickerController.alloc.init
    imagePicker.modalPresentationStyle = UIModalPresentationCurrentContext
    imagePicker.sourceType = UIImagePickerControllerSourceTypePhotoLibrary
    imagePicker.delegate = self

    self.presentViewController(imagePicker, animated:true, completion:nil)

  end

  def clearBackground

    @appearance.backgroundImage = nil
    updateAppearance

  end

  def fontScaleChanged

    @appearance.fontScale = @fontScaleSlider.value
    updateAppearance

  end

  def toggleBoldFont

    if @appearance.fontWeight == "UltraLight"
      @appearance.fontWeight = "DemiBold"
    else
      @appearance.fontWeight = "UltraLight"
    end

    updateAppearance

  end

  def fontBackgroundToggle

    if @fontBackgroundControl.selectedSegmentIndex == 0
      @pickColorOf = "textColor"
      @fontView.alpha = 1
      @backgroundView.alpha = 0
    else
      @pickColorOf = "backgroundColor"
      @fontView.alpha = 0
      @backgroundView.alpha = 1
    end

  end

  def updateAppearance

    @previewView.backgroundColor = @appearance.backgroundColor
    @previewImageView.image = @appearance.backgroundImage
    @previewLabel.setFont(UIFont.fontWithName(@appearance.fontName, size:(@appearance.fontScale * @previewLabel.frame.size.width)))
    @previewLabel.textColor = @appearance.textColor

    if @appearance.fontWeight == "UltraLight"
      @fontWeightButton.setTitle("Bold", forState: UIControlStateNormal)
    else
      @fontWeightButton.setTitle("Light", forState: UIControlStateNormal)
    end

    presentingViewController.updateAppearance if UIDevice.currentDevice.userInterfaceIdiom == UIUserInterfaceIdiomPad
    presentingViewController.outputVC.updateAppearance if presentingViewController.outputVC

  end


  #### image picker deleagte methods ####

  def imagePickerController(picker, didFinishPickingMediaWithInfo:info)

    image = info[UIImagePickerControllerOriginalImage]

    @appearance.backgroundImage = image
    updateAppearance
    
    self.dismissViewControllerAnimated(true, completion:nil)

  end

  def imagePickerControllerDidCancel(picker)

    self.dismissViewControllerAnimated(true, completion:nil)

  end

end

File: Countout/AppearanceController.swift
import UIKit

class AppearanceController: UIViewController {
    
    let appearance = Appearance.sharedClient
    
    var colorPicker: STColorPicker!
    var pickColorOf = ColorOf.Text
    
    enum ColorOf {
        case Text
        case Background
    }

    @IBOutlet weak var countdownView: CountdownView!
    
    @IBOutlet weak var borderView: UIView!
    @IBOutlet weak var settingsView: UIView!
    
    @IBOutlet var fontView: UIView!
    @IBOutlet weak var fontScaleSlider: UISlider!
    @IBOutlet weak var fontWeightButton: UIButton!
    
    @IBOutlet var backgroundView: UIView!
    
    @IBOutlet weak var fontBackgroundControl: UISegmentedControl!
    @IBOutlet weak var colorPickerBorder: UIView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        countdownView.text = String(CountdownTimer.sharedClient.length / 60) + ":" + String(format: "%02d", (CountdownTimer.sharedClient.length % 60))
        
        borderView.backgroundColor = UIColor.clearColor()
        borderView.layer.cornerRadius = 6.0
        borderView.layer.masksToBounds = true
        borderView.layer.borderWidth = 1.0
        borderView.layer.borderColor = UIColor.lightGrayColor().CGColor
        
        NSBundle.mainBundle().loadNibNamed("FontAppearanceView", owner: self, options: nil)
        fontScaleSlider.value = Float(appearance.fontScale)
        borderView.addSubview(fontView)
        fontView.constrainToMatchView(settingsView)
        
        NSBundle.mainBundle().loadNibNamed("BackgroundAppearanceView", owner: self, options: nil)
        backgroundView.alpha = 0
        borderView.addSubview(backgroundView)
        backgroundView.constrainToMatchView(settingsView)
        
        colorPickerBorder.backgroundColor = UIColor.clearColor()
        colorPickerBorder.layer.cornerRadius = 6.0
        colorPickerBorder.layer.masksToBounds = true
        colorPickerBorder.layer.borderWidth = 1.0
        colorPickerBorder.layer.borderColor = UIColor.darkGrayColor().CGColor
        
        colorPicker = STColorPicker(frame: colorPickerBorder.bounds)
        colorPicker.colorHasChanged = { (color: UIColor!, location: CGPoint) in
            if self.pickColorOf == .Text { self.appearance.textColor = color } else { self.appearance.backgroundColor = color }
            self.updateAppearance()
        }
        
        colorPickerBorder.addSubview(colorPicker)
        
        updateAppearance()
    }
    
    override func viewDidLayoutSubviews() {
        colorPicker.frame = colorPickerBorder.bounds
        colorPicker.setColor(self.pickColorOf == .Text ? appearance.textColor : appearance.backgroundColor)
    }
    
    @IBAction func done() {
        if UIDevice.currentDevice().userInterfaceIdiom == .Phone {
            if let countdownController = presentingViewController as? CountdownController {
                countdownController.updateAppearance()
            }
        }
        
        presentingViewController?.dismissViewControllerAnimated(true, completion: nil)
    }
    
    @IBAction func reset() {
        appearance.reset()
        fontScaleSlider.value = Float(appearance.fontScale)
        
        updateAppearance()
    }
    
    @IBAction func chooseBackground() {
        let imagePicker = UIImagePickerController()
        imagePicker.modalPresentationStyle = .CurrentContext
        imagePicker.sourceType = .PhotoLibrary
        imagePicker.delegate = self
        
        self.presentViewController(imagePicker, animated: true, completion: nil)
    }
    
    @IBAction func clearBackground() {
        appearance.backgroundImage = nil
        updateAppearance()
    }
    
    @IBAction func fontScaleChanged() {
        appearance.fontScale = CGFloat(fontScaleSlider.value)
        updateAppearance()
    }
    
    @IBAction func toggleBoldFont() {
        appearance.fontWeight = appearance.fontWeight == "UltraLight" ? "DemiBold" : "UltraLight"
        updateAppearance()
    }
    
    @IBAction func fontBackgroundToggle() {
        if fontBackgroundControl.selectedSegmentIndex == 0 {
            pickColorOf = .Text
            colorPicker.setColor(appearance.textColor)
            fontView.alpha = 1
            backgroundView.alpha = 0
        } else {
            pickColorOf = .Background
            colorPicker.setColor(appearance.backgroundColor)
            fontView.alpha = 0
            backgroundView.alpha = 1
        }
    }
    
    func updateAppearance() {
        countdownView.backgroundColor = appearance.backgroundColor
        countdownView.backgroundImage = appearance.backgroundImage
        countdownView.setFont(name: appearance.fontName, size: appearance.fontScale)
        countdownView.textColor = appearance.textColor
        
        fontWeightButton.setTitle(appearance.fontWeight == "UltraLight" ? "Bold" : "Light", forState: .Normal)
        
        if let countdownController = presentingViewController as? CountdownController {
            if UIDevice.currentDevice().userInterfaceIdiom == .Pad { countdownController.updateAppearance() }
            countdownController.outputVC?.updateAppearance()
        }
    }
}

// MARK: UIImagePickerControllerDelegate
extension AppearanceController: UIImagePickerControllerDelegate {
    func imagePickerController(picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [NSObject : AnyObject]) {
        if let image = info[UIImagePickerControllerOriginalImage] as? UIImage {
            appearance.backgroundImage = image
            updateAppearance()
        }
        
        dismissViewControllerAnimated(true, completion: nil)
    }
    
    func imagePickerControllerDidCancel(picker: UIImagePickerController) {
        dismissViewControllerAnimated(true, completion: nil)
    }
}

// MARK: UINavigationControllerDelegate
// needed for UIImagePickerControllerDelegate???
extension AppearanceController: UINavigationControllerDelegate {
    
}
